File: test/vaccines.test.ts
import { env, SELF } from "cloudflare:test";
import { beforeAll, describe, expect, it } from "vitest";
import { schema } from "../src/worker/db";
import { familyScope } from "../src/worker/db/scoped";
import {
  addMember,
  api,
  createBaby,
  createFamily,
  createUser,
  db,
  rig,
  setPlan,
  signIn,
} from "./helpers";

const BASE = "http://localhost";

// Multipart upload goes straight through SELF: the `api` helper always sends
// a JSON content-type.
async function upload(
  vaccineId: string,
  cookie: string,
  file: { name: string; type: string; bytes: Uint8Array },
) {
  const form = new FormData();
  form.append(
    "file",
    new File([file.bytes as BufferSource], file.name, { type: file.type }),
  );
  return SELF.fetch(`${BASE}/api/vaccines/${vaccineId}/documents`, {
    method: "POST",
    headers: { origin: BASE, cookie },
    body: form,
  });
}

const png = () => new Uint8Array([0x89, 0x50, 0x4e, 0x47, 1, 2, 3, 4]);

async function makeVaccine(cookie: string, babyId: string, name = "MMR") {
  const res = await api("/api/vaccines", {
    method: "POST",
    cookie,
    body: {
      babyId,
      time: new Date().toISOString(),
      name,
      doseNumber: 1,
      scheduleSlot: "mmr:1",
    },
  });
  return (await res.json()) as {
    id: string;
    name: string;
    doseNumber: number | null;
    scheduleSlot: string | null;
    documents: { id: string; url: string }[];
  };
}

describe("vaccine log", () => {
  it("is free to create — no premium gate on the record itself", async () => {
    const { cookie, baby } = await rig();

    const res = await api("/api/vaccines", {
      method: "POST",
      cookie,
      body: {
        babyId: baby.id,
        time: new Date().toISOString(),
        name: "DTP-IPV-Hib-HepB",
        doseNumber: 1,
        scheduleSlot: "dtp-ipv-hib-hepb:1",
      },
    });
    expect(res.status).toBe(201);
    const created = (await res.json()) as {
      name: string;
      scheduleSlot: string | null;
      documents: unknown[];
    };
    expect(created.name).toBe("DTP-IPV-Hib-HepB");
    expect(created.scheduleSlot).toBe("dtp-ipv-hib-hepb:1");
    expect(created.documents).toEqual([]);
  });

  it("accepts an off-programme vaccine with no slot", async () => {
    const { cookie, baby } = await rig();
    const created = await makeVaccine(cookie, baby.id, "Yellow fever");
    const res = await api(`/api/vaccines/${created.id}`, {
      method: "PATCH",
      cookie,
      body: { scheduleSlot: null, doseNumber: null },
    });
    expect(res.status).toBe(200);
    const updated = (await res.json()) as {
      scheduleSlot: string | null;
      doseNumber: number | null;
    };
    expect(updated.scheduleSlot).toBeNull();
    expect(updated.doseNumber).toBeNull();
  });

  it("appears in the timeline under the other filter", async () => {
    const { cookie, baby } = await rig();
    await makeVaccine(cookie, baby.id);

    const res = await api(`/api/timeline?babyId=${baby.id}&filter=other`, {
      cookie,
    });
    const { entries } = (await res.json()) as {
      entries: { kind: string; name?: string }[];
    };
    expect(entries.find((e) => e.kind === "vaccine")?.name).toBe("MMR");
  });

  it("404s on an unknown baby and keeps families apart", async () => {
    const { cookie } = await rig();
    const otherFamily = await createFamily("Other family");
    const theirBaby = await createBaby(otherFamily.id, "Their baby");

    const res = await api("/api/vaccines", {
      method: "POST",
      cookie,
      body: {
        babyId: theirBaby.id,
        time: new Date().toISOString(),
        name: "MMR",
      },
    });
    expect(res.status).toBe(404);
  });
});

describe("vaccine documents", () => {
  it("requires premium to upload, but never to read or delete", async () => {
    const { family, cookie, baby } = await rig();
    const entry = await makeVaccine(cookie, baby.id);

    const denied = await upload(entry.id, cookie, {
      name: "card.png",
      type: "image/png",
      bytes: png(),
    });
    expect(denied.status).toBe(402);
    expect(((await denied.json()) as { code: string }).code).toBe(
      "PLAN_REQUIRED",
    );

    await setPlan(family.id, "premium");
    const ok = await upload(entry.id, cookie, {
      name: "card.png",
      type: "image/png",
      bytes: png(),
    });
    expect(ok.status).toBe(201);
    const doc = (await ok.json()) as { id: string; url: string };
    expect(doc.url).toBe(`/api/files/${doc.id}`);

    // Downgrade: the file must stay readable and deletable.
    await setPlan(family.id, "free");
    const fetched = await api(`/api/files/${doc.id}`, { cookie });
    expect(fetched.status).toBe(200);
    expect(fetched.headers.get("content-type")).toBe("image/png");
    expect(new Uint8Array(await fetched.arrayBuffer())).toEqual(png());

    const removed = await api(`/api/files/${doc.id}`, {
      method: "DELETE",
      cookie,
    });
    expect(removed.status).toBe(200);
    expect((await api(`/api/files/${doc.id}`, { cookie })).status).toBe(404);
  });

  it("serves attachments as downloads, never inline", async () => {
    const { family, cookie, baby } = await rig();
    await setPlan(family.id, "premium");
    const entry = await makeVaccine(cookie, baby.id);
    const res = await upload(entry.id, cookie, {
      name: "card.png",
      type: "image/png",
      bytes: png(),
    });
    const doc = (await res.json()) as { id: string };

    const fetched = await api(`/api/files/${doc.id}`, { cookie });
    expect(fetched.headers.get("content-disposition")).toContain("attachment");
    expect(fetched.headers.get("x-content-type-options")).toBe("nosniff");
  });

  it("refuses types outside the images-and-PDF allowlist", async () => {
    const { family, cookie, baby } = await rig();
    await setPlan(family.id, "premium");
    const entry = await makeVaccine(cookie, baby.id);

    const res = await upload(entry.id, cookie, {
      name: "evil.html",
      type: "text/html",
      bytes: new TextEncoder().encode("<script>alert(1)</script>"),
    });
    expect(res.status).toBe(415);
    expect(((await res.json()) as { code: string }).code).toBe("BAD_TYPE");
  });

  it("refuses a file over the size cap", async () => {
    const { family, cookie, baby } = await rig();
    await setPlan(family.id, "premium");
    const entry = await makeVaccine(cookie, baby.id);

    const res = await upload(entry.id, cookie, {
      name: "huge.pdf",
      type: "application/pdf",
      bytes: new Uint8Array(10 * 1024 * 1024 + 1),
    });
    expect(res.status).toBe(413);
  });

  it("caps the number of files per entry", async () => {
    const { family, cookie, baby } = await rig();
    await setPlan(family.id, "premium");
    const entry = await makeVaccine(cookie, baby.id);

    for (let i = 0; i < 5; i++) {
      const res = await upload(entry.id, cookie, {
        name: `card${i}.png`,
        type: "image/png",
        bytes: png(),
      });
      expect(res.status).toBe(201);
    }
    const sixth = await upload(entry.id, cookie, {
      name: "card5.png",
      type: "image/png",
      bytes: png(),
    });
    expect(sixth.status).toBe(400);
    expect(((await sixth.json()) as { code: string }).code).toBe("TOO_MANY");
  });

  it("hydrates documents onto the entry", async () => {
    const { family, cookie, baby } = await rig();
    await setPlan(family.id, "premium");
    const entry = await makeVaccine(cookie, baby.id);
    await upload(entry.id, cookie, {
      name: "card.png",
      type: "image/png",
      bytes: png(),
    });

    const listed = await api(`/api/vaccines?babyId=${baby.id}`, { cookie });
    const rows = (await listed.json()) as {
      documents: { filename: string; size: number; url: string }[];
    }[];
    expect(rows[0]!.documents).toHaveLength(1);
    expect(rows[0]!.documents[0]!.filename).toBe("card.png");
    expect(rows[0]!.documents[0]!.size).toBe(png().length);
  });

  it("removes the R2 object when the entry is deleted", async () => {
    const { family, cookie, baby } = await rig();
    await setPlan(family.id, "premium");
    const entry = await makeVaccine(cookie, baby.id);
    await upload(entry.id, cookie, {
      name: "card.png",
      type: "image/png",
      bytes: png(),
    });

    const fam = familyScope(db(), family.id);
    const before = await fam.getVaccine(entry.id);
    const key = (await db().select().from(schema.vaccineDocument)).find(
      (d) => d.vaccineLogId === entry.id,
    )!.objectKey;
    expect(before!.documents).toHaveLength(1);
    expect(await env.FILES.get(key)).not.toBeNull();

    const removed = await api(`/api/vaccines/${entry.id}`, {
      method: "DELETE",
      cookie,
    });
    expect(removed.status).toBe(200);
    // No orphan bytes left behind in the bucket.
    expect(await env.FILES.get(key)).toBeNull();
  });

  it("never serves another family's document", async () => {
    const { family, cookie, baby } = await rig();
    await setPlan(family.id, "premium");
    const entry = await makeVaccine(cookie, baby.id);
    const uploaded = await upload(entry.id, cookie, {
      name: "card.png",
      type: "image/png",
      bytes: png(),
    });
    const doc = (await uploaded.json()) as { id: string };

    const outsider = await createUser("Outsider");
    const otherFamily = await createFamily("Other family");
    await addMember(outsider.id, otherFamily.id, "admin");
    const otherCookie = await signIn(outsider.email);

    expect(
      (await api(`/api/files/${doc.id}`, { cookie: otherCookie })).status,
    ).toBe(404);
    expect(
      (
        await api(`/api/files/${doc.id}`, {
          method: "DELETE",
          cookie: otherCookie,
        })
      ).status,
    ).toBe(404);
    // And ours still works, i.e. the refusal above deleted nothing.
    expect((await api(`/api/files/${doc.id}`, { cookie })).status).toBe(200);
  });

  it("refuses to attach to another family's vaccine entry", async () => {
    const { family, cookie, baby } = await rig();
    await setPlan(family.id, "premium");
    const entry = await makeVaccine(cookie, baby.id);

    const outsider = await createUser("Outsider");
    const otherFamily = await createFamily("Other family");
    await addMember(outsider.id, otherFamily.id, "admin");
    await setPlan(otherFamily.id, "premium");
    const otherCookie = await signIn(outsider.email);

    const res = await upload(entry.id, otherCookie, {
      name: "card.png",
      type: "image/png",
      bytes: png(),
    });
    expect(res.status).toBe(404);
  });
});

describe("vaccine dismissals", () => {
  // One sign-in for the whole block: the KV brute-force brake caps sign-ins
  // per IP, and a file that logs in per test eventually trips it. Each test
  // gets its own baby instead, which is the isolation that actually matters
  // here — dismissals are per-baby.
  let shared: Awaited<ReturnType<typeof rig>>;
  beforeAll(async () => {
    shared = await rig("Dismissal family");
  });
  const freshBaby = (name: string) => createBaby(shared.family.id, name);

  it("dismisses a slot, lists it, and restores it — all on the free plan", async () => {
    const { cookie } = shared;
    const baby = await freshBaby("Dismiss baby");

    const created = await api("/api/vaccines/dismissals", {
      method: "POST",
      cookie,
      body: { babyId: baby.id, slotKey: "hpv:1" },
    });
    expect(created.status).toBe(201);
    const dismissal = (await created.json()) as {
      id: string;
      slotKey: string;
    };
    expect(dismissal.slotKey).toBe("hpv:1");

    const listed = await api(`/api/vaccines/dismissals?babyId=${baby.id}`, {
      cookie,
    });
    expect((await listed.json()) as unknown[]).toHaveLength(1);

    const restored = await api(`/api/vaccines/dismissals/${dismissal.id}`, {
      method: "DELETE",
      cookie,
    });
    expect(restored.status).toBe(200);
    const after = await api(`/api/vaccines/dismissals?babyId=${baby.id}`, {
      cookie,
    });
    expect((await after.json()) as unknown[]).toEqual([]);
  });

  it("is idempotent — dismissing twice returns the same row", async () => {
    const { cookie } = shared;
    const baby = await freshBaby("Idempotent baby");
    const body = { babyId: baby.id, slotKey: "mmr:1" };

    const first = await api("/api/vaccines/dismissals", {
      method: "POST",
      cookie,
      body,
    });
    const second = await api("/api/vaccines/dismissals", {
      method: "POST",
      cookie,
      body,
    });
    expect(second.status).toBe(201);
    expect(((await second.json()) as { id: string }).id).toBe(
      ((await first.json()) as { id: string }).id,
    );

    const listed = await api(`/api/vaccines/dismissals?babyId=${baby.id}`, {
      cookie,
    });
    expect((await listed.json()) as unknown[]).toHaveLength(1);
  });

  it("keeps dismissals per baby, not per family", async () => {
    const { cookie } = shared;
    const baby = await freshBaby("Has dismissal");
    const sibling = await freshBaby("Sibling");
    await api("/api/vaccines/dismissals", {
      method: "POST",
      cookie,
      body: { babyId: baby.id, slotKey: "mmr:1" },
    });

    const theirs = await api(`/api/vaccines/dismissals?babyId=${sibling.id}`, {
      cookie,
    });
    expect((await theirs.json()) as unknown[]).toEqual([]);
  });

  it("does not block logging the vaccine afterwards", async () => {
    const { cookie } = shared;
    const baby = await freshBaby("Logs anyway");
    await api("/api/vaccines/dismissals", {
      method: "POST",
      cookie,
      body: { babyId: baby.id, slotKey: "mmr:1" },
    });

    const logged = await api("/api/vaccines", {
      method: "POST",
      cookie,
      body: {
        babyId: baby.id,
        time: new Date().toISOString(),
        name: "MMR",
        doseNumber: 1,
        scheduleSlot: "mmr:1",
      },
    });
    expect(logged.status).toBe(201);
  });

  it("404s on an unknown baby", async () => {
    const { cookie } = shared;
    const otherFamily = await createFamily("Other dismissal family");
    const theirBaby = await createBaby(otherFamily.id, "Their baby");

    const foreign = await api("/api/vaccines/dismissals", {
      method: "POST",
      cookie,
      body: { babyId: theirBaby.id, slotKey: "mmr:1" },
    });
    expect(foreign.status).toBe(404);
  });

  it("never lets another family restore our dismissal", async () => {
    const { cookie } = shared;
    const baby = await freshBaby("Guarded baby");
    const ours = await api("/api/vaccines/dismissals", {
      method: "POST",
      cookie,
      body: { babyId: baby.id, slotKey: "mmr:1" },
    });
    const { id } = (await ours.json()) as { id: string };

    const outsider = await createUser("Dismissal outsider");
    const otherFamily = await createFamily("Outsider family");
    await addMember(outsider.id, otherFamily.id, "admin");
    const otherCookie = await signIn(outsider.email);

    expect(
      (
        await api(`/api/vaccines/dismissals/${id}`, {
          method: "DELETE",
          cookie: otherCookie,
        })
      ).status,
    ).toBe(404);
    // Still ours, i.e. the refusal deleted nothing.
    const listed = await api(`/api/vaccines/dismissals?babyId=${baby.id}`, {
      cookie,
    });
    expect((await listed.json()) as unknown[]).toHaveLength(1);
  });

  it("does not mistake the dismissals path for a vaccine id", async () => {
    const { cookie } = shared;
    const baby = await freshBaby("Routing baby");
    // Ordering hazard: /api/vaccines/{id} could swallow "dismissals".
    const res = await api(`/api/vaccines/dismissals?babyId=${baby.id}`, {
      cookie,
    });
    expect(res.status).toBe(200);
    expect(Array.isArray(await res.json())).toBe(true);
  });
});

File: apps/server/internal/api/vaccines_test.go
package api_test

import (
	"net/http"
	"testing"
	"time"

	"github.com/refsdal/pjokk/server/internal/testrig"
)

// -----------------------------------------------------------------------
// Vaccine log — ports apps/api/test/vaccines.test.ts's "vaccine log"
// describe block, minus the timeline-integration test (no /api/timeline
// route on this port yet).
// -----------------------------------------------------------------------

func TestCreateVaccineIsFreeAndDocumentsStartEmpty(t *testing.T) {
	a := testrig.App(t)
	familyID, cookie := a.NewFamily("Hansen", "parent@example.com")
	babyID := a.NewBaby(familyID, "Nora")

	res := a.Do(http.MethodPost, "/api/vaccines", cookie, map[string]any{
		"babyId":       babyID,
		"time":         time.Now().UTC().Format(time.RFC3339),
		"name":         "DTP-IPV-Hib-HepB",
		"doseNumber":   1,
		"scheduleSlot": "dtp-ipv-hib-hepb:1",
	})
	if res.Status != http.StatusCreated {
		t.Fatalf("status = %d, body %s", res.Status, res.Raw)
	}
	if res.JSON["name"] != "DTP-IPV-Hib-HepB" {
		t.Errorf("name = %v, want %q", res.JSON["name"], "DTP-IPV-Hib-HepB")
	}
	if res.JSON["scheduleSlot"] != "dtp-ipv-hib-hepb:1" {
		t.Errorf("scheduleSlot = %v, want %q", res.JSON["scheduleSlot"], "dtp-ipv-hib-hepb:1")
	}
	docs, ok := res.JSON["documents"].([]any)
	if !ok || len(docs) != 0 {
		t.Errorf("documents = %v, want an empty array", res.JSON["documents"])
	}
}

func TestUpdateVaccineAcceptsOffProgrammeClears(t *testing.T) {
	a := testrig.App(t)
	familyID, cookie := a.NewFamily("Hansen", "parent@example.com")
	babyID := a.NewBaby(familyID, "Nora")

	created := a.Do(http.MethodPost, "/api/vaccines", cookie, map[string]any{
		"babyId":       babyID,
		"time":         time.Now().UTC().Format(time.RFC3339),
		"name":         "Yellow fever",
		"doseNumber":   1,
		"scheduleSlot": "mmr:1",
	})
	if created.Status != http.StatusCreated {
		t.Fatalf("POST status = %d, body %s", created.Status, created.Raw)
	}
	id, _ := created.JSON["id"].(string)

	updated := a.Do(http.MethodPatch, "/api/vaccines/"+id, cookie, map[string]any{
		"scheduleSlot": nil,
		"doseNumber":   nil,
	})
	if updated.Status != http.StatusOK {
		t.Fatalf("PATCH status = %d, body %s", updated.Status, updated.Raw)
	}
	if updated.JSON["scheduleSlot"] != nil {
		t.Errorf("scheduleSlot after clear = %v, want nil", updated.JSON["scheduleSlot"])
	}
	if updated.JSON["doseNumber"] != nil {
		t.Errorf("doseNumber after clear = %v, want nil", updated.JSON["doseNumber"])
	}
	if updated.JSON["name"] != "Yellow fever" {
		t.Errorf("name after unrelated update = %v, want unchanged %q", updated.JSON["name"], "Yellow fever")
	}
}

func TestCreateVaccineUnknownBabyIs404(t *testing.T) {
	a := testrig.App(t)
	_, cookieA := a.NewFamily("Hansen", "parent@example.com")
	familyB, _ := a.NewFamily("Other family", "other@example.com")
	theirBaby := a.NewBaby(familyB, "Their baby")

	res := a.Do(http.MethodPost, "/api/vaccines", cookieA, map[string]any{
		"babyId": theirBaby,
		"time":   time.Now().UTC().Format(time.RFC3339),
		"name":   "MMR",
	})
	if res.Status != http.StatusNotFound {
		t.Fatalf("status = %d, body %s, want 404", res.Status, res.Raw)
	}
	if res.JSON["error"] != "Unknown baby" || res.JSON["code"] != "NOT_FOUND" {
		t.Errorf("body = %v, want {error:\"Unknown baby\",code:\"NOT_FOUND\"}", res.JSON)
	}
}

// -----------------------------------------------------------------------
// Family scoping / listing / delete.
// -----------------------------------------------------------------------

func TestListVaccinesNewestFirstWithCaretakerName(t *testing.T) {
	a := testrig.App(t)
	familyID, cookie := a.NewFamily("Hansen", "parent@example.com")
	babyID := a.NewBaby(familyID, "Nora")

	t0 := time.Date(2026, 1, 1, 12, 0, 0, 0, time.UTC)
	first := a.Do(http.MethodPost, "/api/vaccines", cookie, map[string]any{
		"babyId": babyID, "time": t0.Format(time.RFC3339), "name": "MMR",
	})
	if first.Status != http.StatusCreated {
		t.Fatalf("first POST status = %d, body %s", first.Status, first.Raw)
	}
	if first.JSON["caretakerName"] != "Rig admin" {
		t.Errorf("caretakerName = %v, want %q", first.JSON["caretakerName"], "Rig admin")
	}
	second := a.Do(http.MethodPost, "/api/vaccines", cookie, map[string]any{
		"babyId": babyID, "time": t0.Add(time.Hour).Format(time.RFC3339), "name": "HPV",
	})
	if second.Status != http.StatusCreated {
		t.Fatalf("second POST status = %d, body %s", second.Status, second.Raw)
	}

	list := a.DoArray(http.MethodGet, "/api/vaccines?babyId="+babyID, cookie, nil)
	if len(list.JSON) != 2 {
		t.Fatalf("GET /api/vaccines = %v, want 2", list.JSON)
	}
	newest := list.JSON[0].(map[string]any)
	if newest["id"] != second.JSON["id"] {
		t.Errorf("newest-first: first row id = %v, want the later entry %v", newest["id"], second.JSON["id"])
	}
}

func TestVaccinesAreFamilyScoped(t *testing.T) {
	a := testrig.App(t)
	familyA, cookieA := a.NewFamily("Family A", "a@example.com")
	_, cookieB := a.NewFamily("Family B", "b@example.com")
	babyA := a.NewBaby(familyA, "Baby A")

	created := a.Do(http.MethodPost, "/api/vaccines", cookieA, map[string]any{
		"babyId": babyA, "time": time.Now().UTC().Format(time.RFC3339), "name": "MMR",
	})
	if created.Status != http.StatusCreated {
		t.Fatalf("POST status = %d, body %s", created.Status, created.Raw)
	}
	id, _ := created.JSON["id"].(string)

	listA := a.DoArray(http.MethodGet, "/api/vaccines", cookieA, nil)
	listB := a.DoArray(http.MethodGet, "/api/vaccines", cookieB, nil)
	if len(listA.JSON) != 1 {
		t.Errorf("family A list = %v, want 1", listA.JSON)
	}
	if len(listB.JSON) != 0 {
		t.Errorf("family B list = %v, want 0", listB.JSON)
	}

	patch := a.Do(http.MethodPatch, "/api/vaccines/"+id, cookieB, map[string]any{"notes": "nope"})
	if patch.Status != http.StatusNotFound {
		t.Errorf("cross-family PATCH status = %d, want 404", patch.Status)
	}
	del := a.Do(http.MethodDelete, "/api/vaccines/"+id, cookieB, nil)
	if del.Status != http.StatusNotFound {
		t.Errorf("cross-family DELETE status = %d, want 404", del.Status)
	}
}

func TestDeleteVaccineRemovesItAndUnknownIDIs404(t *testing.T) {
	a := testrig.App(t)
	familyID, cookie := a.NewFamily("Hansen", "parent@example.com")
	babyID := a.NewBaby(familyID, "Nora")

	created := a.Do(http.MethodPost, "/api/vaccines", cookie, map[string]any{
		"babyId": babyID, "time": time.Now().UTC().Format(time.RFC3339), "name": "MMR",
	})
	id, _ := created.JSON["id"].(string)

	del := a.Do(http.MethodDelete, "/api/vaccines/"+id, cookie, nil)
	if del.Status != http.StatusOK {
		t.Fatalf("DELETE status = %d, body %s", del.Status, del.Raw)
	}
	if del.JSON["ok"] != true {
		t.Errorf("DELETE body = %v, want {ok:true}", del.JSON)
	}

	again := a.Do(http.MethodDelete, "/api/vaccines/"+id, cookie, nil)
	if again.Status != http.StatusNotFound {
		t.Fatalf("second DELETE status = %d, body %s, want 404", again.Status, again.Raw)
	}
}

func TestVaccinesRejectsUnauthenticated(t *testing.T) {
	a := testrig.App(t)
	res := a.Do(http.MethodGet, "/api/vaccines", "", nil)
	if res.Status != http.StatusUnauthorized {
		t.Fatalf("status = %d, want 401", res.Status)
	}
}

// -----------------------------------------------------------------------
// Vaccine dismissals — ports apps/api/test/vaccines.test.ts's "vaccine
// dismissals" describe block.
// -----------------------------------------------------------------------

func TestDismissRestoreVaccineSlot(t *testing.T) {
	a := testrig.App(t)
	familyID, cookie := a.NewFamily("Dismissal family", "parent@example.com")
	babyID := a.NewBaby(familyID, "Dismiss baby")

	created := a.Do(http.MethodPost, "/api/vaccines/dismissals", cookie, map[string]any{
		"babyId": babyID, "slotKey": "hpv:1",
	})
	if created.Status != http.StatusCreated {
		t.Fatalf("POST status = %d, body %s", created.Status, created.Raw)
	}
	if created.JSON["slotKey"] != "hpv:1" {
		t.Errorf("slotKey = %v, want %q", created.JSON["slotKey"], "hpv:1")
	}
	id, _ := created.JSON["id"].(string)

	listed := a.DoArray(http.MethodGet, "/api/vaccines/dismissals?babyId="+babyID, cookie, nil)
	if len(listed.JSON) != 1 {
		t.Fatalf("listed = %v, want 1", listed.JSON)
	}

	restored := a.Do(http.MethodDelete, "/api/vaccines/dismissals/"+id, cookie, nil)
	if restored.Status != http.StatusOK {
		t.Fatalf("DELETE status = %d, body %s", restored.Status, restored.Raw)
	}

	after := a.DoArray(http.MethodGet, "/api/vaccines/dismissals?babyId="+babyID, cookie, nil)
	if len(after.JSON) != 0 {
		t.Errorf("after restore = %v, want an empty array", after.JSON)
	}
}

func TestCreateVaccineDismissalIsIdempotent(t *testing.T) {
	a := testrig.App(t)
	familyID, cookie := a.NewFamily("Dismissal family", "parent@example.com")
	babyID := a.NewBaby(familyID, "Idempotent baby")
	body := map[string]any{"babyId": babyID, "slotKey": "mmr:1"}

	first := a.Do(http.MethodPost, "/api/vaccines/dismissals", cookie, body)
	if first.Status != http.StatusCreated {
		t.Fatalf("first POST status = %d, body %s", first.Status, first.Raw)
	}
	second := a.Do(http.MethodPost, "/api/vaccines/dismissals", cookie, body)
	if second.Status != http.StatusCreated {
		t.Fatalf("second POST status = %d, body %s, want 201", second.Status, second.Raw)
	}
	if second.JSON["id"] != first.JSON["id"] {
		t.Errorf("second create id = %v, want the same row %v", second.JSON["id"], first.JSON["id"])
	}

	listed := a.DoArray(http.MethodGet, "/api/vaccines/dismissals?babyId="+babyID, cookie, nil)
	if len(listed.JSON) != 1 {
		t.Errorf("listed = %v, want 1 (idempotent)", listed.JSON)
	}
}

func TestVaccineDismissalsArePerBabyNotPerFamily(t *testing.T) {
	a := testrig.App(t)
	familyID, cookie := a.NewFamily("Dismissal family", "parent@example.com")
	babyID := a.NewBaby(familyID, "Has dismissal")
	sibling := a.NewBaby(familyID, "Sibling")

	if res := a.Do(http.MethodPost, "/api/vaccines/dismissals", cookie, map[string]any{
		"babyId": babyID, "slotKey": "mmr:1",
	}); res.Status != http.StatusCreated {
		t.Fatalf("POST status = %d, body %s", res.Status, res.Raw)
	}

	theirs := a.DoArray(http.MethodGet, "/api/vaccines/dismissals?babyId="+sibling, cookie, nil)
	if len(theirs.JSON) != 0 {
		t.Errorf("sibling's dismissals = %v, want an empty array", theirs.JSON)
	}
}

func TestVaccineDismissalDoesNotBlockLoggingTheVaccine(t *testing.T) {
	a := testrig.App(t)
	familyID, cookie := a.NewFamily("Dismissal family", "parent@example.com")
	babyID := a.NewBaby(familyID, "Logs anyway")

	if res := a.Do(http.MethodPost, "/api/vaccines/dismissals", cookie, map[string]any{
		"babyId": babyID, "slotKey": "mmr:1",
	}); res.Status != http.StatusCreated {
		t.Fatalf("dismissal POST status = %d, body %s", res.Status, res.Raw)
	}

	logged := a.Do(http.MethodPost, "/api/vaccines", cookie, map[string]any{
		"babyId":       babyID,
		"time":         time.Now().UTC().Format(time.RFC3339),
		"name":         "MMR",
		"doseNumber":   1,
		"scheduleSlot": "mmr:1",
	})
	if logged.Status != http.StatusCreated {
		t.Errorf("vaccine POST status = %d, body %s, want 201", logged.Status, logged.Raw)
	}
}

func TestCreateVaccineDismissalUnknownBabyIs404(t *testing.T) {
	a := testrig.App(t)
	_, cookieA := a.NewFamily("Dismissal family", "parent@example.com")
	familyB, _ := a.NewFamily("Other dismissal family", "other@example.com")
	theirBaby := a.NewBaby(familyB, "Their baby")

	res := a.Do(http.MethodPost, "/api/vaccines/dismissals", cookieA, map[string]any{
		"babyId": theirBaby, "slotKey": "mmr:1",
	})
	if res.Status != http.StatusNotFound {
		t.Fatalf("status = %d, body %s, want 404", res.Status, res.Raw)
	}
}

func TestVaccineDismissalIsFamilyScoped(t *testing.T) {
	a := testrig.App(t)
	familyID, cookie := a.NewFamily("Dismissal family", "parent@example.com")
	babyID := a.NewBaby(familyID, "Guarded baby")
	_, otherCookie := a.NewFamily("Outsider family", "outsider@example.com")

	created := a.Do(http.MethodPost, "/api/vaccines/dismissals", cookie, map[string]any{
		"babyId": babyID, "slotKey": "mmr:1",
	})
	if created.Status != http.StatusCreated {
		t.Fatalf("POST status = %d, body %s", created.Status, created.Raw)
	}
	id, _ := created.JSON["id"].(string)

	del := a.Do(http.MethodDelete, "/api/vaccines/dismissals/"+id, otherCookie, nil)
	if del.Status != http.StatusNotFound {
		t.Errorf("cross-family DELETE status = %d, want 404", del.Status)
	}
	// Still ours — the refusal deleted nothing.
	listed := a.DoArray(http.MethodGet, "/api/vaccines/dismissals?babyId="+babyID, cookie, nil)
	if len(listed.JSON) != 1 {
		t.Errorf("listed after refused cross-family delete = %v, want 1", listed.JSON)
	}
}

// Ordering hazard: /api/vaccines/{id} could swallow "dismissals" as an id.
// Go's net/http.ServeMux prefers the literal "/api/vaccines/dismissals"
// pattern over the "/api/vaccines/{id}" wildcard regardless of
// registration order, but this proves it end to end rather than trusting
// that guarantee blind.
func TestVaccineDismissalsPathIsNotCapturedAsAnId(t *testing.T) {
	a := testrig.App(t)
	familyID, cookie := a.NewFamily("Routing family", "parent@example.com")
	babyID := a.NewBaby(familyID, "Routing baby")

	res := a.DoArray(http.MethodGet, "/api/vaccines/dismissals?babyId="+babyID, cookie, nil)
	if res.Status != http.StatusOK {
		t.Fatalf("status = %d, body %s, want 200", res.Status, res.Raw)
	}
}
